### PR TITLE
[IBCDPE-572] Add in contribution guide

### PR DESCRIPTION
**Pull request by thomasyu888**
contribution guide for the repository
* adding tests
* developing locally
* etc

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+## Coding Style
+
+The code in this package is also automatically formatted by `black` for consistency.
+
+## The Development Life Cycle
+
+### Fork and clone this repository
+
+1. See the [Github docs](https://help.github.com/articles/fork-a-repo/) for how to make a copy (a fork) of a repository to your own Github account.
```
Comment (JessterB) [inline CONTRIBUTING.md:15]: I thought we weren't working on forks anymore in this repo?

Comment (thomasyu888) [reply CONTRIBUTING.md:15]: Well, this section would be if people outside of Sage developed, but.. we can remove just to simplify this.

Comment (JessterB) [reply CONTRIBUTING.md:15]: iirc, CI will fail if we get a PR from a fork because the fork doesn't have access to the required secrets (or something along those lines).

**Review by BWMac — APPROVED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (jaclynbeck-sage) [reply CONTRIBUTING.md:15]: That's correct, CI won't be able to log into synapse if the PR is from a fork because of access to secrets. I'm not sure how to get around that for external contributors...

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+1. Then, [clone the repository](https://help.github.com/articles/cloning-a-repository/) to your local machine so you can begin making changes.
+1. Add this repository as an [upstream remote](https://help.github.com/en/articles/configuring-a-remote-for-a-fork) on your local git repository so that you are able to fetch the latest commits.
+1. On your local machine make sure you have the latest version of the `develop` branch:
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md:18]: Isn't the branch called `dev`, not `develop`?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+> To ensure the most fluid development, do not push to your `dev`
+
+1. Access
+    1. (Assuming you have followed all 4 steps above in the "fork and clone this repository" section). Navigate to your cloned repository on your computer/server. 
+    1. If you are Sage Bio employee working directly on this project, please ask for write permissions to contribute to the repository.
+1. Make sure you are always creating feature branches from the `dev` branch.
+
+    ```shell
+    # On a fork
+    cd {your-github-username}/Genie
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: I don't understand what `Genie` is or why the forked repo would be named with `{your-github-username}`. Wouldn't it just be `agora-data-tools`?
I don't really understand what this section of code is doing. Is `develop` a branch? Or is this supposed to mean `dev`? Why are we checking it out twice?

Comment (thomasyu888) [reply CONTRIBUTING.md]: Thanks!  This is the result of poor copy pasta review.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+Please follow the [README.md](README.md) to install the package for development purposes.
+
+### Developing at Sage Bio
+
+The agora-data-tools project follows the standard [trunk based development](https://trunkbaseddevelopment.com/) development strategy.
+> To ensure the most fluid development, do not push to your `dev`
+
+1. Access
+    1. (Assuming you have followed all 4 steps above in the "fork and clone this repository" section). Navigate to your cloned repository on your computer/server. 
+    1. If you are Sage Bio employee working directly on this project, please ask for write permissions to contribute to the repository.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Maybe reword as "please ask for write permissions to contribute directly to this repository, rather than creating a fork"?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    # On a fork
+    cd {your-github-username}/Genie
+    git checkout develop
+    git pull upstream develop
+    # Write access
+    git checkout develop
+    git pull
+    ```
+
+1. Create a feature branch which off the `dev` branch. If there is JIRA issue that you are addressing, name the branch after the issue with some more detail (like `{user}/{GH|JIRA}-123/add-some-new-feature`).
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Typo: "which off the dev branch".

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    ```shell
+    git commit changed_file.txt -m "Remove X parameter because it was unused"
+    git push
+    ```
+
+1. (Make sure you have follow instructions in "Install development dependencies") Once you have made your additions or changes, make sure you write tests and run the test suite.  More information on testing below.
+
+    ```shell
+    pytest ./test
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Why is this `pytest ./test` but below it's `pytest -vs tests/`? I think the first one (here) wouldn't work?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+1. Always merge all new features into `develop` branch first (unless it is a documentation, readme, or github action patch into `main`)
+1. After initial features are ready in the `develop` branch, create a `release-X.X` branch to prepare for the release.
+    1. update the `__version__` parameter in `genie/__init__.py`
+1. Merge `release-X.X` branch into `main` - Not by pull request!
+1. Create release tag (`v...`) and include release notes.  Also include any known bugs for each release here.
+1. Merge `main` back into `develop` -->
+
+### Transforms
+
+This package has a `src/agoradatatools/etl/transform` submodule.  This folder houses all the individual transform modules required for the package.  Here are the steps to add more transforms
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Needs punctuation at the end.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+1. Create new script in transform submodule that matches the dataset name and name the function `transform_...`
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: should be "in _the_ transform submodule".

Also might be useful to give an example. Like, "For example, if you have a dataset named `genome_variants`, your new script would be `src/agoradatatools/etl/transform/transform_genome_variants.py`."

File: CONTRIBUTING.md (hunk 1 above)
Comment (thomasyu888) [reply CONTRIBUTING.md:15]: What we could do is just explicitly state the contribution guidelines are specific to sage employees that work on this. 

Once we figure out a "better" way of doing post processing validation, we may not need to have the synapse login.

File: CONTRIBUTING.md (hunk 4 above)
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Missing word: "If you are _a_ Sage Bio employee"

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+It is recommended to use the following style (see example below) for mock testing across this package:
+
+```python
+from unittest.mock import patch
+...
+patch.object(MODULE_NAME, "FUNCTION_TO_MOCK_NAME".return_value=SOME_RETURN_VALUE)
+```
+
+<!-- ### Release Procedure (For Package Maintainers)
```
Comment (JessterB) [inline CONTRIBUTING.md:114]: Are all the commented out bits targeted at package maintainers (or some other specific) audience? If so, it might be good to mention that up front either globally if all the same audience, or per chunk if not.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    - These tests should include multiple ways of evaluating the transform function, including one test that should pass (good input data) and at least one that should fail (bad input data).
+    - For some functions, it may be appropriate to include multiple passing datasets (e.g. for functions that are written to handle imperfect data) and/or multiple failing datasets (e.g. for transforms operating on datasets that can be unclean in multiple distinct ways).
+    - Each transform function should have its own folder in `test_assets` to hold its input and output data files. Inputs should be in CSV form and outputs in JSON form.
+    - Use `pytest.mark.parameterize` to loop through multiple datasets in a single test.
+    - The class `TestTransformGenesBiodomains` can be used as an example for future tests contibuted.
+
+### DockerHub
+
+This repository does not use github actions to push docker images.  By adding the `sagebiodockerhub` github user as an Admin to this GitHub repository, we can configure an automated build in DockerHub.  You can view the repo [here](https://hub.docker.com/r/sagebionetworks/agora-data-tools).
```
Comment (JessterB) [inline CONTRIBUTING.md]: This confused me a bit, but I think it's saying:

Rather than using github actions to build and push docker images to DockerHub, the docker images are automatically built on DockerHub. This requires that the `sagebiodockerhub` user is an Admin of this repo.

Did I interpret that correctly?

**Review by JessterB — APPROVED** [review]
I suggested a couple of small clarifications, it looks good to me overall!

File: CONTRIBUTING.md (hunk 9 above)
Comment (thomasyu888) [reply CONTRIBUTING.md:114]: I think its a bit confusing when you are reading the raw markdown, but you won't see it when you click on the rendered markdown: https://github.com/Sage-Bionetworks/agora-data-tools/blob/tyu/add-contributing/CONTRIBUTING.md.

This section is commented out, because it hasn't been fully established.

File: CONTRIBUTING.md (hunk 10 above)
Comment (thomasyu888) [reply CONTRIBUTING.md]: Yes, that is the correct interpretation - I'd actually like us to switch to use gh-actions to build to gh-packages so that everything in contained in the repo.  I'll reword

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    ``` -->
+
+### Install development dependencies
+
+Please follow the [README.md](README.md) to install the package for development purposes.
+
+### Developing at Sage Bio
+
+The agora-data-tools project follows the standard [trunk based development](https://trunkbaseddevelopment.com/) development strategy.
+> To ensure the most fluid development, do not push to your `dev`
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Since we've removed instructions on forking, I think this should just be "do not push to `dev`" (without 'your').

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+1. Please ask for write permissions to contribute directly to this repository.
+1. Make sure you are always creating feature branches from the `dev` branch.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: It might be good to include a note that CI will fail for pull requests from forks of the repo, which is why we do branches.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    ```
+
+    You should now be able to see the branch on GitHub. Make commits as you deem necessary. It helps to provide useful commit messages - a commit message saying 'Update' is a lot less helpful than saying 'Remove X parameter because it was unused'.
+
+    ```shell
+    git commit changed_file.txt -m "Remove X parameter because it was unused"
+    git push
+    ```
+
+1. (Make sure you have follow instructions in "Install development dependencies") Once you have made your additions or changes, make sure you write tests and run the test suite.  More information on testing below.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Typo: follow**ed** instructions.
Can we link to the relevant page/section that corresponds to "Install development dependencies"? Is this in the Readme.md file? Also it doesn't make sense why this comment is down here right before the section on testing, and not up at the top with the Readme info.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+    ```shell
+    pytest -vs tests/
+    ```
+
+1. (Make sure you have follow instructions in "Install development dependencies") Make sure to run the auto python code formatter, black.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Same as above.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+    ```shell
+    black ./
+    ```
+
+1. Once you have completed all the steps above, create a pull request from the feature branch to the `dev` branch of the Sage-Bionetworks/agora-data-tools repo.
+
+> *A code maintainer must review and accept your pull request.* A code review ideally happens with both the contributor and the reviewer present, but is not strictly required for contributing. This can be performed remotely (e.g., Zoom, Hangout, or other video or phone conference).  Most code reviews are done asyncronously.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: I'm a little confused by the the mention of contributor and reviewer being present, because we actually _don't_ do reviews via Zoom etc at all. Or at least, Jess and I don't? Plus we usually have more than one reviewer.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+This package uses [semantic versioning](https://semver.org/) for releasing new versions. A github release should occur at least once a quarter to capture the changes between releases.
```
Comment (jaclynbeck-sage) [inline CONTRIBUTING.md]: Who is "responsible" for updating the version? I see a commented out section below, so if this is still TBD it's fine to leave this as-is for now.

**Review by jaclynbeck-sage — COMMENTED** [review]
Mostly nit-picks at this point, from me trying to view this as a "new" employee who hasn't seen how things work. The directions on development lifecycle and testing are really clear and easy to follow, I appreciate that you put shell commands!

File: CONTRIBUTING.md (hunk 16 above)
Comment (thomasyu888) [reply CONTRIBUTING.md]: This is a TBD.  For now, I am minting the releases when I feel "appropriate", but thats just more of a "feeling"

I personally like more frequent releases to reduce the size of the changelog, I can add that.

**Review by jaclynbeck-sage — APPROVED** [review]
Everything looks great! Thanks for the updates.
